community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/BadCollector.java

```diff
@@ -56,16 +68,44 @@ public BadCollector( OutputStream out, int tolerance, int collect )
     }
 
     @Override
-    public void collectBadRelationship( InputRelationship relationship, Object specificValue )
+    public void collectBadRelationship( final InputRelationship relationship, final Object specificValue )
     {
-        checkTolerance( BAD_RELATIONSHIPS,
-                new InputException( format( "%s refering to missing node %s", relationship, specificValue ) ) );
+        checkTolerance( BAD_RELATIONSHIPS, new Bad()
```

Comment:
It's the new `Bad` instance vs. creating an Exception, which is much more expensive (as profiling also shows). Bad defers creating exception until we exceed tolerance. We still need the message to log though.